windows/devices/uwb/UwbConnector.cxx

```diff
@@ -149,33 +151,40 @@ UwbConnector::GetCapabilities()
         return resultFuture;
     }
 
-    // Determine the amount of memory required for the UWB_DEVICE_CAPABILITIES from the driver.
-    DWORD bytesRequired = 0;
-    BOOL ioResult = DeviceIoControl(handleDriver.get(), IOCTL_UWB_GET_DEVICE_CAPABILITIES, nullptr, 0, nullptr, 0, &bytesRequired, nullptr);
-    if (!LOG_IF_WIN32_BOOL_FALSE(ioResult)) {
-        // TODO: need to do something different here
-        HRESULT hr = HRESULT_FROM_WIN32(GetLastError());
-        PLOG_ERROR << "error when sending IOCTL_UWB_GET_DEVICE_CAPABILITIES, hr=" << std::showbase << std::hex << hr;
-        return resultFuture;
-    }
+    std::vector<uint8_t> uwbDeviceCapabilitiesBuffer{};
+    DWORD bytesRequired = offsetof(UWB_DEVICE_CAPABILITIES, capabilityParams);
 
-    // Allocate memory for the UWB_DEVICE_CAPABILITIES structure, and pass this to the driver request.
-    DWORD uwbCapabilitiesSize = bytesRequired;
-    auto uwbDeviceCapabilitiesBuffer = std::make_unique<uint8_t[]>(uwbCapabilitiesSize);
-    ioResult = DeviceIoControl(handleDriver.get(), IOCTL_UWB_GET_DEVICE_CAPABILITIES, nullptr, 0, uwbDeviceCapabilitiesBuffer.get(), uwbCapabilitiesSize, &bytesRequired, nullptr);
-    if (!ioResult) {
-        // TODO: need to do something different here
-        HRESULT hr = HRESULT_FROM_WIN32(GetLastError());
-        PLOG_ERROR << "error when sending IOCTL_UWB_GET_DEVICE_CAPABILITIES, hr=" << std::showbase << std::hex << hr;
-        return resultFuture;
+    for (const auto i : std::ranges::iota_view{ 0, 2 }) {
+        uwbDeviceCapabilitiesBuffer.resize(bytesRequired);
+        PLOG_DEBUG << "IOCTL_UWB_GET_DEVICE_CAPABILITIES attempt #" << (i + 1) << " with " << std::size(uwbDeviceCapabilitiesBuffer) << "-byte buffer";
+        BOOL ioResult = DeviceIoControl(handleDriver.get(), IOCTL_UWB_GET_DEVICE_CAPABILITIES, nullptr, 0, std::data(uwbDeviceCapabilitiesBuffer), std::size(uwbDeviceCapabilitiesBuffer), &bytesRequired, nullptr);
+        if (!LOG_IF_WIN32_BOOL_FALSE(ioResult)) {
+            DWORD lastError = GetLastError();
+            // Treat all errors other than insufficient buffer size as fatal.
+            if (lastError != ERROR_MORE_DATA) {
+                HRESULT hr = HRESULT_FROM_WIN32(lastError);
+                PLOG_ERROR << "error when sending IOCTL_UWB_GET_DEVICE_CAPABILITIES, hr=" << std::showbase << std::hex << hr;
+                resultPromise.set_exception(std::make_exception_ptr(UwbException(UwbStatusGeneric::Failed)));
+                break;
+            }
+            // Attempt to retry the ioctl with the appropriate buffer size, which is now held in bytesRequired.
+            const auto& deviceCapabilitiesPartial = *reinterpret_cast<UWB_DEVICE_CAPABILITIES*>(std::data(uwbDeviceCapabilitiesBuffer));
+            bytesRequired = deviceCapabilitiesPartial.size;

```

Comment:
One thing I forgot to do in the pre-existing impl was to validate that `bytesRequired >= offsetof(UWB_DEVICE_CAPABILITIES, size)`. Without that, we'd be accessing invalid memory if the driver didn't return enough data. Not likely to happen, but we need to code defensively against it.

Can you please add that here and to the existing ones? (I think `IOCTL_UWB_NOTIFICATION` and `IOCTL_UWB_SET_APP_CONFIG_PARAMS`).

Reply:
I don't quite understand the need for this. If the `size` field is later in the struct, how would that affect the buffer size? It's the same value right?

Reply:
Here, the amount of data returned by the ioctl is `bytesRequired`. So, before we access stuff inside `deviceCapabilitiesPartial`, we need to ensure that we're accessing a part that's actually been filled in by the ioctl. Since we're accessing the `size` field, we know it's at offset `offsetof(UWB_DEVICE_CAPABILITIES, size)` and it has size `sizeof(UWB_DEVICE_CAPABILITIES::size)`, so, `bytesRequired` has to be >= `offsetof(UWB_DEVICE_CAPABILITIES, size) + offsetof(UWB_DEVICE_CAPABILITIES, size)`, otherwise we'd be accessing memory past the end of the ioctl output buffer. My main point was that using this expression for calculating the validity of accessing the `size` member doesn't make any assumptions about where `size` is in the partially filled structure. Does that make more sense?

Reply:
I see, I understand the reasoning now. However, it's not clear where this check would happen. The amount of data returned by the IOCTL isn't in `bytesRequired`, we set `bytesRequired = deviceInformationPartial.size`, so the check would need to happen after this. But this still results in the same problem in the case where the `size` field is at the end of the struct and the buffer is too small to contain that field - we'd be accessing invalid memory when reading `deviceInformationPartial.size`. Right?

Reply:
I see, I understand the reasoning now. However, it's not clear where this check would happen. The amount of data returned by the IOCTL isn't in `bytesRequired` right after the `DeviceIoControl` call, the amount of data would go in the `size` field, then we set `bytesRequired = deviceInformationPartial.size`. So the check would need to happen after this, but this still results in the same problem in the case where the `size` field is at the end of the struct and the buffer is too small to contain that field - we'd be accessing invalid memory when reading `deviceInformationPartial.size`. Right?

Reply:
You're right, yeah. This is a dilemma since this approach is inconsistent with the `DeviceIoControl` contract which states this:

![image](https://user-images.githubusercontent.com/2082148/231000458-65d876fa-105b-4c8d-ad36-18abbea63a72.png)

That's exactly the case here: `ERROR_GET_MORE_DATA` was returned and the `DeviceIoContrtol` API tells us that the amount of data returned is in `lpBytesReturned`, which is what one would expect and is intuitive.

Let's leave it as-is for now. I'd like to think about it some more and see if there's a way to make this consistent, intuitive, and doesn't break the `DeviceIoControl` contract.